### PR TITLE
Add support for partial http requests handling with pipelining

### PR DESCRIPTION
**Pull request by mhl-b**
UPDATE: 
Created a leaner and cleaner version here https://github.com/elastic/elasticsearch/pull/111438. 
Most of changes here are irrelevant in new PR.

-----

This PR adds support to `Netty4HttpPipeliningHandler` to handle parts of HTTP request. Right now we always aggregate HTTP request content before passing to `RestController`. With this change `Netty4HttpPipeliningHandler` can receive `FullHttpRequests` and `HttpRequests`/`HttpContents` with correct sequence number.

To make it possible I made several changes in Netty channel pipeline.
1. Add HTTP pipeline sequence to parts of HTTP request. Before `Netty4HttpPipeliningHandler` would keep track of all incoming `FullHttpRequests` and increase sequence number. With this change addition of sequence number happens sooner, after `HttpRequestDecoder`, but before decompression and aggregation. It happens in `Netty4InboundHttpPipeliningHandler`.

2. Propagation of sequence number through `HttpObjectAggregator` and `HttpContentDecompressor`. Since netty does not know about our pipelining implementation, I have to wrap both classes that can consume and produce a "pipelined" versions of `HttpObjects`. Besides wrapper handler classes I introduced strong-typed versions of `HttpRequest, HttpContent, LastHttpContent, FullHttpRequest` with pipeline sequence number, for example 
`public record PipelinedHttpContent(HttpContent httpContent, int sequence)`
3. Make HTTP aggregation optional. In this PR we still do full aggregation to all requests. But I added a predicate that can control which requests can skip aggregation, and unit tests.

After change pipeline looks like this:
```
| from                        | to                                 | note                       |
|-----------------------------+------------------------------------+----------------------------|
| HttpRequestDecoder          | HttpRequestDecoder                 |                            |
| Netty4HttpHeaderValidator   | Netty4HttpHeaderValidator          |                            |
|                             | Netty4InboundHttpPipeliningHandler | emits PipelinedHttpObjects |
| HttpContentDecompressor     | Netty4ContentDecompressor          | wrapped decompressor       |
| HttpObjectAggregator        | Netty4HttpAggregator               | optional aggregation       |
| Netty4HttpPipeliningHandler | Netty4HttpPipeliningHandler        |                            |
```

Most interesting changes in `Netty4ContentDecompressor, Netty4HttpAggregator, Netty4InboundHttpPipeliningHandler , Netty4HttpPipeliningHandler` and related unit tests at the end. The rest is boilerplate.

**Update**:
Added example of RestHandler with chunked content. I use [reactive streams model](https://www.reactive-streams.org/). Rather than exposing `BytesReference` as content, I use `Flow.Publisher<HttpContent>`. The RestHandler will subscribe to publisher and consume parts at it's own rate - by [`Subscription.request(long n)`](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/util/concurrent/Flow.Subscription.html#request(long)). The publisher will read from netty channel on demand and invoke `onNext(HttpContent)` when number of requested parts is greater than 0.

Content publisher - `Netty4RequestContentPublisher`
Rest handler example - `Netty4RequestContentPublisherIT`

I omitted error handling and edge cases for brevity. There many things to consider with publisher state machine: for example avoid multiple subscriptions, handling terminal states, handling errors, handling subscriber cancellation. I did some drafts, and I can say it's pretty easy to add them.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java [hunk 1]
```diff
@@ -109,28 +107,40 @@ public Netty4HttpPipeliningHandler(
...
+                nonError = new Exception(cause);
+            } else {
+                nonError = (Exception) cause;
+            }
+            netty4HttpRequest = new Netty4HttpRequest(request.sequence(), request, nonError);
+        } else {
+            netty4HttpRequest = new Netty4HttpRequest(request.sequence(), request);
+        }
+        handlePipelinedRequest(ctx, netty4HttpRequest);
+    }
```
Comment (mhl-b) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:138]: no changes here, moved into separate method

**Review by DaveCTurner — COMMENTED** [review]
Thanks Mikhail. I'll take a more detailed look in due course but two things straight away:

1. Could we have a test case which shows how a REST handler would use this API? Probably a `ESNetty4IntegTestCase` derivative would be best, see e.g. `Netty4ChunkedContinuationsIT` for an example of the sort of thing I mean.

2. Could we merge this into a feature branch for now rather than `main`? That way we've somewhere to coordinate the work to adopt this in the bulk indexing path, but we're not committing to it until we have the whole solution in place and can run some benchmarks?

**Comment by mhl-b** [comment]
@DaveCTurner 
1. Make sense, will do
2. Feature branch - done

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/PipelinedHttpRequestPart.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.http.netty4;
+
+/**
+ * A part of HTTP request that has HTTP pipelining sequence number
+ */
+public sealed interface PipelinedHttpRequestPart extends PipelinedHttpObject permits PipelinedHttpRequest, PipelinedHttpContent,
+    PipelinedLastHttpContent {
```
Comment (ywangd) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/PipelinedHttpRequestPart.java:15]: I'd prefer to drop this marker interface. IIUC, it is only to differentiate `PipelinedFullHttpRequest` from other `PipelinedHttpObject`. Can we instead have a method on `PipelinedHttpObject` to differentiate between them? I personally find the hierarchy a bit hard to follow.

**Review by ywangd — COMMENTED** [review]
This seems to be an exciting piece of work! Thanks for working on it. I had a first read-through and left some minor comments. I plan to come back to it and have a closer look. In the meantime, I second on David comment about having a test case that demonstrates how this new feature is leveraged.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/PipelinedHttpRequestPart.java (hunk 2 above)
Comment (mhl-b) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/PipelinedHttpRequestPart.java:15]: I planned to use this sealed interface in PartialRequest path, so I can do `switch(part)` with exhaustive check. Which is part of java 21 [jep-441](https://openjdk.org/jeps/441). But we are still compiling with java 17 and nothing fancy came out of this code. Can be removed.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4ContentDecompressor.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * A wrapper around {@link HttpContentDecompressor} that consumes and produces {@link PipelinedHttpRequest} and
+ * {@link PipelinedHttpContent}.
+ */
+public class Netty4ContentDecompressor extends HttpContentDecompressor {
+
+    @Override
+    public boolean acceptInboundMessage(Object msg) throws Exception {
+        return msg instanceof PipelinedHttpObject;
```
[the remark this answers is not in the document]
Comment (mhl-b) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4ContentDecompressor.java:29]: Wrapped Decompressor should call this method on every "read(ctx, msg)", I added this check to avoid surprises in "decode" method that assumes message has sequence number.

**Comment by mhl-b** [comment]
@DaveCTurner @ywangd 
I updated PR description with approach and implementation of RestHandler with chunked request content.
Please see `Netty4RequestContentPublisherIT` for an API design. TLDR I use [reactive streams](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/util/concurrent/Flow.html).

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * A wrapper around {@link HttpObjectAggregator}, provides optional aggregation for {@link PipelinedHttpObject}'s.
+ * A {@code decider} predicate selects HTTP requests that will be aggregated into {@link PipelinedFullHttpRequest}.
+ */
+public class Netty4HttpAggregator extends HttpObjectAggregator {
+
+    private static final Predicate<PipelinedHttpRequest> IGNORE_TEST = (req) -> req.uri().startsWith("/_test/request-stream") == false;
```
Comment (mhl-b) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java:25]: added to the source for the sake of PR size

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java [hunk 5]
```diff
@@ -109,23 +112,34 @@ public Netty4HttpPipeliningHandler(
...
+                    }
+                    netty4HttpRequest = new Netty4HttpRequest(sequence, (FullHttpRequest) request, nonError);
                 } else {
-                    nonError = (Exception) cause;
+                    if (request instanceof FullHttpRequest fullRequest) {
+                        netty4HttpRequest = new Netty4HttpRequest(sequence, fullRequest);
+                    } else {
+                        var contentPublisher = new Netty4RequestContentPublisher(ctx.channel());
+                        netty4HttpRequest = new Netty4HttpRequest(sequence, request, contentPublisher);
+                        currentRequest = netty4HttpRequest;
```
Comment (mhl-b) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:135]: note: I attach current request to Handler object and following chunks will go to the same object until we see next request

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java [hunk 6]
```diff
@@ -109,23 +112,34 @@ public Netty4HttpPipeliningHandler(
...
+                    }
                 }
-                netty4HttpRequest = new Netty4HttpRequest(readSequence++, fullHttpRequest, nonError);
+                handlePipelinedRequest(ctx, netty4HttpRequest);
             } else {
-                netty4HttpRequest = new Netty4HttpRequest(readSequence++, fullHttpRequest);
+                assert currentRequest != null;
+                currentRequest.contentPublisher().sendChunk((HttpContent) msg);
```
Comment (mhl-b) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:141]: I attach currentRequest to the Netty4PipeliningHandler and consequent chunks can be published right a way to the RestHandler from here.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java [hunk 7]
```diff
@@ -36,39 +42,64 @@
...
+        this.sequence = sequence;
+        this.nettyRequest = request;
+        this.nettyContent = new DefaultHttpContent(Unpooled.EMPTY_BUFFER);
+        this.contentPublisher = contentPublisher;
+        this.content = BytesArray.EMPTY;
+        this.released = new AtomicBoolean(false);
+        this.pooled = false;
+        this.headers = getHttpHeadersAsMap(request.headers());
+        this.inboundException = null;
+    }
```
Comment (mhl-b) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:67]: Add support to Netty4HttpRequest to be HttpRequest and optional HttpContent, rather than always FullHttpRequest. I do drop trailing headers for not full requests, not sure do we use them or not. Haven't found traces of them yet.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4RequestContentPublisher.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+package org.elasticsearch.http.netty4;
+
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.LastHttpContent;
+
+import org.elasticsearch.http.HttpContent;
+
+import java.util.concurrent.Flow;
+
+public class Netty4RequestContentPublisher implements Flow.Publisher<HttpContent> {
```
Comment (mhl-b) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4RequestContentPublisher.java:18]: bare bones, nothing fancy, no error handling, no state management, no multiple subscriptions, just happy path

**Review by DaveCTurner — COMMENTED** [review]
A reactive stream is the right conceptual model for sure but I'm not convinced we want to stick to these exact APIs. There's always exactly one subscriber, known ahead of time, so there's no need for a separate `subscribe()` API; we really want `.request()` to request a certain number of bytes but will typically deliver this as a single chunk, which differs from the reactive streams API that expects the argument to be the number of chunks; finally the `Throwable` in the `onError` path is very concerning. Remember this code is all on an incredibly hot path, we need to avoid unnecessary abstractions wherever possible.

**Comment by mhl-b** [comment]
@DaveCTurner @ywangd 
I created new PR https://github.com/elastic/elasticsearch/pull/111438. 
I found simpler and smaller implementation, that does not have problems introduced here. No changes in pipelining sequencing, no more PipelinedObjects, simplified wrapper for HttpObjectAggregator that does not intercept "decode" method, simplified interface for streamed content, I still use reactive streams as a model but only with minimal required methods.

**Comment by ywangd** [comment]
Would it be OK to close this PR if this is not intended to be continued?
